raiden-ts/src/transfers/mediate/types.ts

```diff
@@ -54,44 +65,290 @@ export interface FeeModel<Config, Schedule> {
+ * @param func - Discrete func
+ * @returns Found index
+ */
+function findRangeIndex(x: Decimal, func: DiscreteFunc): number {
+  assert(func.length >= 2, 'invalid linspace');
+  const x0 = func[0][0];
+  const xL = func[func.length - 1][0]; // x_last
+  // special-case where x is exactly over xLast, make it part of last stretch instead of beyond it
+  if (x.eq(xL.toHexString())) return func.length - 2;
+  const width = new Decimal(xL.sub(x0).toHexString());
+  const step = width.div(func.length - 1);
+
+  let index = Math.floor(x.sub(x0.toHexString()).div(step).toNumber());
+  let offs = 0;
+  do {
+    if (index < 0) return -1;
+    else if (index >= func.length - 1) return func.length - 1;
+
+    if (x.gte(func[index + 1][0].toHexString())) offs = -1;
+    else if (x.lt(func[index][0].toHexString())) offs = 1;
+    else offs = 0;
+
+    index += offs;
+  } while (offs);
+  return index;
+}
+
+function interpolate(x: Decimal, [x0, y0]: Point, [x1, y1]: Point): Decimal {
+  // y = f(x) = (x-x0)*Δy/Δx + y0
+  return x
+    .sub(x0.toHexString())
+    .mul(y1.sub(y0).toHexString())
+    .div(x1.sub(x0).toHexString())
+    .add(y0.toHexString());
+}
+
+/**
+ * Get the value of the piecewise linear curve described by the pairs in [arr] at point [x]
+ *
+ * @param x - Point where to get the value for
+ * @param arr - function described as array of [x, y] pairs containing points linked by lines
+ * @returns function value at point x
+ */
+function interpolateFunc(x: Decimal, arr: DiscreteFunc): Decimal {
+  let index = findRangeIndex(x, arr);
+  // in case index before|beyond arr limits, use first|last stretches
+  if (index < 0) index = 0;
+  else if (index >= arr.length - 1) index = arr.length - 2;
+  return interpolate(x, arr[index], arr[index + 1]);
+}
+
+/**
+ * Calculates pair of points discretizing a U-shaped curve which describes channel's imbalance fees
+ * for a given total capacity and config.
+ *
+ * @param channelCapacity - Channel's total capacity (sum of deposits from both ends)
+ * @param proportionalImbalanceFee - Fee config, in PPM (1% = 10,000)
+ * @returns array of [x, y] pairs, where x is the old/new channel capacity and y is the fee
+ */
+function calculatePenaltyFunction(
+  channelCapacity: UInt<32>,
+  proportionalImbalanceFee: number,
+): DiscreteFunc<UInt<32>> {
+  const NUM_DISCRETISATION_POINTS = 21;
+  assert(channelCapacity.gt(0), 'not enough capacity');
+
+  const bnChannelCapacity = new Decimal(channelCapacity.toHexString());
+  const numBasePoints = Decimal.min(
+    NUM_DISCRETISATION_POINTS,
+    bnChannelCapacity.add(1),
+  ).toNumber();
+  const xValues = linspace(Zero as UInt<32>, channelCapacity, numBasePoints);
+
+  if (proportionalImbalanceFee === 0) {
+    return [
+      [Zero as UInt<32>, Zero as UInt<32>],
+      [channelCapacity as UInt<32>, Zero as UInt<32>],
+    ];
+  }
+  const bnProportionalImbalanceFee = new Decimal(proportionalImbalanceFee).div(1e6);
+
+  const MAX_SLOPE = new Decimal('0.1');
```

Comment:
nit: Please move that to the beginning of the function

Reply:
Sure. It went here because I replicated PC's respective function, but I'll move. Thanks